### PR TITLE
Portable Ruby 3.3.1

### PR DESCRIPTION
**Pull request by MikeMcQuaid**
Use the latest version of Portable Ruby.

I may have missed some stuff here.

May need new checksums once https://github.com/Homebrew/homebrew-portable-ruby/actions/runs/9105311684 completes.

### CODE REVIEW

**Comment by Bo98** [comment]
If we're deleting 3.1 vendored gems may need to bump HOMEBREW_REQUIRED_RUBY_VERSION, which is technically a breaking change but let's pretend we didn't miss 4.3.0.

File: .gitignore [hunk 1]
```diff
@@ -40,7 +40,7 @@
 # Ignore Ruby gems for versions other than we explicitly vendor.
 # Keep this in sync with the list in standalone/init.rb.
```
Comment (Bo98) [inline .gitignore:41]: This needs updating

**Comment by MikeMcQuaid** [comment]
> If we're deleting 3.1 vendored gems may need to bump HOMEBREW_REQUIRED_RUBY_VERSION, which is technically a breaking change but let's pretend we didn't miss 4.3.0.

Yeh, might just retrospectively add to the notes there if I can get this merged 🔜 and note it'll land in 4.3.1.

File: Library/Homebrew/cask/utils.rb [hunk 2]
```diff
@@ -47,7 +47,7 @@ def self.gain_permissions_remove(path, command: SystemCommand)
       gain_permissions(path, permission_flags, command) do |p|
         if p.parent.writable?
           if directory
-            p.rmtree
+            FileUtils.rm_r p
```
Comment (Bo98) [inline Library/Homebrew/cask/utils.rb:50]: `Pathname#rmtree` now never returning errors is a horrifically subtle change that questions the safety of every `#rmtree` call we have, given the new behaviour now means `rmtree` can not actually remove the directory.

Comment (Bo98) [reply Library/Homebrew/cask/utils.rb:50]: e.g. this error catching now doesn't work:

https://github.com/Homebrew/brew/blob/97030b799262c331ad017f5a20cd2c0b9a13d8d1/Library/Homebrew/reinstall.rb#L73-L78

We could override the behaviour in `extend/pathname.rb`. Seems hacky but might be the safest solution.

**Comment by EricFromCanada** [comment]
Seeing the same test failure in 10.13 and up that was showing in CI earlier:
```
1) Cask::Artifact::App install_phase when the target already exists given the force option target is user-owned but contains read-only files overwrites the existing app
     Failure/Error: let(:install_phase) { app.install_phase(command:, adopt:, force:) }

     ErrorDuringExecution:
       Failure while executing; `/usr/bin/env touch /private/tmp/homebrew-tests-20240516-52284-zi9hyc/cask-appdir/Caffeine.app/.homebrew-write-test` exited with 1. Here's the output:
       touch: /private/tmp/homebrew-tests-20240516-52284-zi9hyc/cask-appdir/Caffeine.app/.homebrew-write-test: Permission denied
```
10.11 and 10.12 weren't able to install the required gems for `brew tests`.

**Comment by Bo98** [comment]
[sorbet/sorbet#7895](https://redirect.github.com/sorbet/sorbet/pull/7895)

**Comment by Bo98** [comment]
> 10.11 and 10.12 weren't able to install the required gems for `brew tests`.

I assume this was the case with Ruby 3.1.4 too?

**Review by Bo98 — APPROVED** [review]

File: Library/Homebrew/cask/utils.rb (hunk 2 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/cask/utils.rb]: Yeh, good catch and idea here @Bo98.

**Comment by MikeMcQuaid** [comment]
Thanks for all the help here @Bo98 🎉

**Pull request merged**